src/LanguageServer/Protocol/Handler/ServerLifetime/LspServiceLifeCycleManager.cs

```diff
@@ -14,22 +18,46 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.ServerLifetime;
 internal sealed class LspServiceLifeCycleManager : ILifeCycleManager, ILspService
 {
     private readonly IClientLanguageServerManager _clientLanguageServerManager;
+    private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService;
 
-    public LspServiceLifeCycleManager(IClientLanguageServerManager clientLanguageServerManager)
+    [ExportCSharpVisualBasicLspServiceFactory(typeof(LspServiceLifeCycleManager)), Shared]
+    [method: ImportingConstructor]
+    [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    internal class LspLifeCycleManagerFactory(LspWorkspaceRegistrationService lspWorkspaceRegistrationService) : ILspServiceFactory
+    {
+        public ILspService CreateILspService(LspServices lspServices, WellKnownLspServerKinds serverKind)
+        {
+            var clientLanguageServerManager = lspServices.GetRequiredService<IClientLanguageServerManager>();
+            return new LspServiceLifeCycleManager(clientLanguageServerManager, lspWorkspaceRegistrationService);
+        }
+    }
+
+    private LspServiceLifeCycleManager(IClientLanguageServerManager clientLanguageServerManager, LspWorkspaceRegistrationService lspWorkspaceRegistrationService)
     {
         _clientLanguageServerManager = clientLanguageServerManager;
+        _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
     }
 
     public async Task ShutdownAsync(string message = "Shutting down")
     {
+        // Shutting down is not cancellable.
+        var cancellationToken = CancellationToken.None;
+
+        var hostWorkspace = _lspWorkspaceRegistrationService.GetAllRegistrations().SingleOrDefault(w => w.Kind == WorkspaceKind.Host);
+        if (hostWorkspace is not null)
+        {
+            var service = hostWorkspace.Services.GetRequiredService<IExtensionMessageHandlerService>();
+            await service.ResetAsync(cancellationToken).ConfigureAwait(false);
+        }
```

Comment:
@dibarbet can you look at this section to make sure htis is ok?

Reply:
this looks fine